common/client/node_lifecycle.go

```diff
@@ -0,0 +1,431 @@
+package client
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+
+	"github.com/smartcontractkit/chainlink/v2/core/utils"
+)
+
+var (
+	promPoolRPCNodeHighestSeenBlock = promauto.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "pool_rpc_node_highest_seen_block",
+		Help: "The highest seen block for the given RPC node",
+	}, []string{"chainID", "nodeName"})
+	promPoolRPCNodeNumSeenBlocks = promauto.NewCounterVec(prometheus.CounterOpts{
+		Name: "pool_rpc_node_num_seen_blocks",
+		Help: "The total number of new blocks seen by the given RPC node",
+	}, []string{"chainID", "nodeName"})
+	promPoolRPCNodePolls = promauto.NewCounterVec(prometheus.CounterOpts{
+		Name: "pool_rpc_node_polls_total",
+		Help: "The total number of poll checks for the given RPC node",
+	}, []string{"chainID", "nodeName"})
+	promPoolRPCNodePollsFailed = promauto.NewCounterVec(prometheus.CounterOpts{
+		Name: "pool_rpc_node_polls_failed",
+		Help: "The total number of failed poll checks for the given RPC node",
+	}, []string{"chainID", "nodeName"})
+	promPoolRPCNodePollsSuccess = promauto.NewCounterVec(prometheus.CounterOpts{
+		Name: "pool_rpc_node_polls_success",
+		Help: "The total number of successful poll checks for the given RPC node",
+	}, []string{"chainID", "nodeName"})
+)
+
+// zombieNodeCheckInterval controls how often to re-check to see if we need to
+// state change in case we have to force a state transition due to no available
+// nodes.
+// NOTE: This only applies to out-of-sync nodes if they are the last available node
+func zombieNodeCheckInterval(noNewHeadsThreshold time.Duration) time.Duration {
+	interval := noNewHeadsThreshold
+	if interval <= 0 || interval > QueryTimeout {
+		interval = QueryTimeout
+	}
+	return utils.WithJitter(interval)
+}
+
+func (n *node[CHAIN_ID, HEAD, RPC]) setLatestReceived(blockNumber int64, totalDifficulty *utils.Big) {
+	n.stateMu.Lock()
+	defer n.stateMu.Unlock()
+	n.stateLatestBlockNumber = blockNumber
+	n.stateLatestTotalDifficulty = totalDifficulty
+}
+
+const (
+	msgCannotDisable = "but cannot disable this connection because there are no other RPC endpoints, or all other RPC endpoints are dead."
+	msgDegradedState = "Chainlink is now operating in a degraded state and urgent action is required to resolve the issue"
+)
+
+// Node is a FSM
+// Each state has a loop that goes with it, which monitors the node and moves it into another state as necessary.
+// Only one loop must run at a time.
+// Each loop passes control onto the next loop as it exits, except when the node is Closed which terminates the loop permanently.
+
+// This handles node lifecycle for the ALIVE state
+// Should only be run ONCE per node, after a successful Dial
+func (n *node[CHAIN_ID, HEAD, RPC]) aliveLoop() {
```

Comment:
This method is quite complex and large it's difficult to operate in such context. Can we extract parts of it into separate methods?

Reply:
Same as above about existing code.